Firestore/core/test/firebase/firestore/util/executor_test.cc

```diff
@@ -243,6 +244,60 @@ TEST_P(ExecutorTest, DuplicateTagsOnOperationsAreAllowed) {
   EXPECT_EQ(steps, "12");
 }
 
+TEST_P(ExecutorTest, ConcurrentExecutorsWork) {
+  /**
+   * A mix of a countdown latch and a barrier. All threads that bump the
+   * countdown block until the count becomes zero.
+   */
+  class BlockingCountdown {
+   public:
+    explicit BlockingCountdown(int threads) : threads_(threads) {
+    }
+
+    int count() const {
+      std::lock_guard<std::mutex> lock(mutex_);
+      return threads_;
+    }
+
+    /** Awaits the completion of all threads. */
+    void Await() {
+      std::unique_lock<std::mutex> lock(mutex_);
+      while (threads_ > 0) {
+        is_zero_.wait(lock);
+      }
+    }
+
+    /**
+     * Bumps the counter down by one and waits for the counter to become zero.
+     */
+    void Bump() {
+      std::unique_lock<std::mutex> lock(mutex_);
+
+      threads_ -= 1;
+
+      is_zero_.wait(lock, [this] { return threads_ == 0; });
+
+      is_zero_.notify_all();
+    }
+
+   private:
+    mutable std::mutex mutex_;
+    std::condition_variable is_zero_;
+    int threads_ = 0;
+  };
+
+  const int threads = 5;
```

Comment:
Would it be worthwhile to check the case where there's more/less tasks than threads? (i.e., check with `4` and `6` threads)

Reply:
Checking with more tasks than there are backing threads would deadlock. Checking with fewer tasks than the number of available threads doesn't really test anything different than what this is testing so I'm disinclined to add the test.